### PR TITLE
DISPATCH-1780: first stab at aggregated multicast

### PR DESCRIPTION
**Pull request by grs**

### CODE REVIEW

File: src/adaptors/http1/http1_client.c [hunk 1]
```diff
@@ -1218,6 +1406,26 @@ uint64_t qdr_http1_client_core_link_deliver(qdr_http1_adaptor_t    *adaptor,
     _client_response_msg_t *rmsg = DEQ_TAIL(hreq->responses);
     assert(rmsg && rmsg->dlv == delivery);
 
+    if (hconn->cfg.aggregation != QD_AGGREGATION_NONE) {
```
Comment (kgiusti) [inline src/adaptors/http1/http1_client.c:1409]: Pls add comment explaining that responses are saved and not encoded until after the request has been accepted.

File: src/adaptors/http1/http1_client.c [hunk 2]
```diff
@@ -541,7 +546,12 @@ static void _client_tx_buffers_cb(h1_codec_request_state_t *hrs, qd_buffer_list_
     // responses are decoded one at a time - the current response it at the
     // tail of the response list
 
-    _client_response_msg_t *rmsg = DEQ_TAIL(hreq->responses);
+    _client_response_msg_t *rmsg;
+    if (hconn->cfg.aggregation == QD_AGGREGATION_NONE) {
+        rmsg = DEQ_TAIL(hreq->responses);
+    } else {
+        rmsg = DEQ_HEAD(hreq->responses);
```
[the remark this answers is not in the document]
Comment (kgiusti) [reply src/adaptors/http1/http1_client.c:553]: Update: now I understand - we buffer all responses and hold off writing them until after request is acked.  Need to update comment line 546.

File: src/adaptors/http1/http1_client.c [hunk 3]
```diff
@@ -575,7 +585,12 @@ static void _client_tx_stream_data_cb(h1_codec_request_state_t *hrs, qd_message_
...
```
[the remark this answers is not in the document]
Comment (kgiusti) [reply src/adaptors/http1/http1_client.c:592]: Brain re-assembled - update comment on line 585.

**Review by kgiusti — APPROVED** [review]
See inline - mostly documentation nits.

File: python/qpid_dispatch/management/qdrouter.json [hunk 4]
```diff
@@ -1124,6 +1124,18 @@
                     "default": "HTTP1",
                     "required": false,
                     "create": true
+                },
+                "aggregation": {
+                    "type": "string",
```
[the remark this answers is not in the document]
Comment (grs) [reply python/qpid_dispatch/management/qdrouter.json:1129]: Will do!

File: src/adaptors/http1/http1_codec.c [hunk 5]
```diff
@@ -1528,6 +1533,62 @@ static inline void _flush_headers(h1_codec_request_state_t *hrs, struct encoder_
     }
 }
 
+int h1_codec_tx_begin_multipart(h1_codec_request_state_t *hrs)
+{
+    h1_codec_connection_t *conn = h1_codec_request_state_get_connection(hrs);
+    struct encoder_t *encoder = &conn->encoder;
+    encoder->boundary_marker = (char*) malloc(QD_DISCRIMINATOR_SIZE + 2);
```
[the remark this answers is not in the document]
Comment (grs) [reply src/adaptors/http1/http1_codec.c:1540]: Oops! yes it should!

File: src/adaptors/http1/http1_client.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (grs) [reply src/adaptors/http1/http1_client.c:553]: I'll update the comment to explain, but yes, when aggregating we don't send out separate reponses for each response message. Using the head for the single aggregated response is arbitrary I suppose but it seemed simplest.

File: src/adaptors/http1/http1_client.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (grs) [reply src/adaptors/http1/http1_client.c:592]: Will do!

File: src/adaptors/http1/http1_client.c (hunk 1 above)
Comment (grs) [reply src/adaptors/http1/http1_client.c:1409]: Will do

File: src/adaptors/http1/http1_client.c [hunk 6]
```diff
@@ -942,6 +1126,10 @@ void qdr_http1_client_core_delivery_update(qdr_http1_adaptor_t      *adaptor,
                 qdr_http1_error_response(&hreq->base, 503, "Service Unavailable");
```
[the remark this answers is not in the document]
Comment (grs) [reply src/adaptors/http1/http1_client.c:1126]: Good point. I think for an aggregated response to a multicast request, the actual outcome of the request is not relevant. I'll fix.

File: src/adaptors/http1/http1_server.c [hunk 7]
```diff
@@ -1056,11 +1063,10 @@ void qdr_http1_server_core_link_flow(qdr_http1_adaptor_t    *adaptor,
                 rmsg->dlv = qdr_link_deliver_to(hconn->in_link, rmsg->msg, 0, addr, false, 0, 0, 0, 0);
                 qdr_delivery_set_context(rmsg->dlv, (void*) hreq);
                 rmsg->msg = 0;
-                if (!rmsg->rx_complete) {
+                if (!rmsg->rx_complete || hconn->cfg.aggregation != QD_AGGREGATION_NONE) {
                     // stop here since response must be complete before we can deliver the next one.
```
[the remark this answers is not in the document]
Comment (grs) [reply src/adaptors/http1/http1_server.c:1067]: Will update comment. I don't want to free the response until it is accepted.

**Pull request closed without merging**